src/yunohost/certificate.py

```diff
@@ -0,0 +1,760 @@
+
+
+def _email_renewing_failed(domain, exception_message, stack):
+    from_ = "certmanager@%s (Certificate Manager)" % domain
+    to_ = "root"
+    subject_ = "Certificate renewing attempt for %s failed!" % domain
+
+    logs = _tail(50, "/var/log/yunohost/yunohost-cli.log")
+    text = """
+An attempt for renewing the certificate for domain %s failed with the following
+error :
+
+%s
+%s
+
+Here's the tail of /var/log/yunohost/yunohost-cli.log, which might help to
+investigate :
+
+%s
+
+-- Certificate Manager
+
+""" % (domain, exception_message, stack, logs)
+
+    message = """
+From: %s
+To: %s
+Subject: %s
+
+%s
+""" % (from_, to_, subject_, text)
+
+    smtp = smtplib.SMTP("localhost")
+    smtp.sendmail(from_, [to_], message)
+    smtp.quit()
+
+
+def _configure_for_acme_challenge(auth, domain):
+    nginx_conf_file = "/etc/nginx/conf.d/%s.d/000-acmechallenge.conf" % domain
+
+    nginx_configuration = '''
+location '/.well-known/acme-challenge'
+{
+        default_type "text/plain";
+        alias %s;
+}
+    ''' % WEBROOT_FOLDER
+
+    # Write the conf
+    if os.path.exists(nginx_conf_file):
+        logger.info("Nginx configuration file for ACME challenge already exists for domain, skipping.")
+        return
+
+    logger.info("Adding Nginx configuration file for Acme challenge for domain %s.", domain)
+
+    with open(nginx_conf_file, "w") as f:
+        f.write(nginx_configuration)
+
+    # Assume nginx conf is okay, and reload it
+    # (FIXME : maybe add a check that it is, using nginx -t, haven't found
+    # any clean function already implemented in yunohost to do this though)
+    _run_service_command("reload", "nginx")
+
+    app_ssowatconf(auth)
+
+
+def _fetch_and_enable_new_certificate(domain):
+    # Make sure tmp folder exists
+    logger.debug("Making sure tmp folders exists...")
+
+    if not os.path.exists(WEBROOT_FOLDER):
+        os.makedirs(WEBROOT_FOLDER)
+
+    if not os.path.exists(TMP_FOLDER):
+        os.makedirs(TMP_FOLDER)
+
+    _set_permissions(WEBROOT_FOLDER, "root", "www-data", 0650)
+    _set_permissions(TMP_FOLDER, "root", "root", 0640)
+
+    # Prepare certificate signing request
+    logger.info("Prepare key and certificate signing request (CSR) for %s...", domain)
+
+    domain_key_file = "%s/%s.pem" % (TMP_FOLDER, domain)
+    _generate_key(domain_key_file)
+    _set_permissions(domain_key_file, "root", "metronome", 0640)
+
+    _prepare_certificate_signing_request(domain, domain_key_file, TMP_FOLDER)
+
+    # Sign the certificate
+    logger.info("Now using ACME Tiny to sign the certificate...")
+
+    domain_csr_file = "%s/%s.csr" % (TMP_FOLDER, domain)
+
+    signed_certificate = sign_certificate(ACCOUNT_KEY_FILE,
+                                          domain_csr_file,
+                                          WEBROOT_FOLDER,
+                                          log=logger,
+                                          CA=CERTIFICATION_AUTHORITY)
```

Comment:
Done, included the call to acme-tiny in a try/catch, with a particular handling for this exception pointing to https://letsencrypt.org/docs/rate-limits/. We might want to catch other exceptions which might occur often, but it's tricky to know this in advance :confused: